Sources/NIOHPACK/HPACKHeader.swift

```diff
@@ -276,42 +265,159 @@ public struct HPACKHeaders: ExpressibleByDictionaryLiteral, Sendable {
+        return self._values(forHeader: name, canonicalForm: true)
+    }
+
+    /// Return a sequence of header values for the header field named `name`.
+    ///
+    /// See also ``subscript(name:)`` and ``subscript(canonicalForm:)``.
+    ///
+    /// - Parameters:
+    ///   - name: The name of the header field whose values should be iterated.
+    ///   - canonicalForm: Whether the values should be decomposed into their "canonical form". That
+    ///     is, splitting them on commas as extensively as possible such that multiple values
+    ///     received on the one line are returned as separate entries.
+    /// - Note: Enabling `canonicalForm`  for the 'set-cookie' header has no efffect.
+    /// - Returns: A sequence of header values for `name`.
+    public func values(forHeader name: String, canonicalForm: Bool = false) -> Values {
+        if canonicalForm, name.isEqualCaseInsensitiveASCIIBytes(to: "set-cookie") {
+            // Not safe to split 'set-cookie'.
+            return Values(headers: self, name: name, canonicalize: false)
+        } else {
+            return Values(headers: self, name: name, canonicalize: canonicalForm)
+        }
+    }
+
+    /// Special internal function for use by tests.
+    internal subscript(position: Int) -> (String, String) {
+        precondition(position < self.headers.endIndex, "Position \(position) is beyond bounds of \(self.headers.endIndex)")
+        let header = self.headers[position]
+        return (header.name, header.value)
+    }
+
+    @inlinable
+    internal func _values(forHeader name: String, canonicalForm: Bool) -> [String] {
+        if self.headers.isEmpty {
             return []
         }
 
-        // It's not safe to split Set-Cookie on comma.
-        if name.isEqualCaseInsensitiveASCIIBytes(to: "set-cookie") {
-            return result
+        // Estimate the size of the array at the cost of a second scan through the header table.
+        // This allows us to avoid the incremental allocations we would otherwise have without
+        // reserving capacity in the array.
+        let count = self._countValues(forHeader: name)
+
+        // Allow for some extra space when getting the canonical form.
+        let minimumCapacity = canonicalForm ? 4 * count : count
+
+        var values = [String]()
+        values.reserveCapacity(minimumCapacity)
+        values.append(contentsOf: self.values(forHeader: name, canonicalForm: canonicalForm))
+        return values
+    }
+
+    @inlinable
+    internal func _countValues(forHeader name: String) -> Int {
+        var count = 0
+        for header in self.headers {
+            if header.name.isEqualCaseInsensitiveASCIIBytes(to: name) {
+                // Cannot overflow: count is no greater than headers.count.
+                count &+= 1
+            }
+        }
+        return count
+    }
+}
+
+extension HPACKHeaders {
+    public struct Values: Sequence {
+        public typealias Element = String
```

Comment:
Given that we're currently trying to reduce allocations, should we take this opportunity to return `Substring` instead of `String`?

Reply:
Yeah, good idea.